Cannot Compile

I see somebody else was running into this same issue on the llm-chain-template repo (https://github.com/sobelio/llm-chain-template/issues/1). Thought it might be useful to post this issue here since it seems to be an issue with the llm-chain-openai crate itself.

I'm getting the following error when trying to build the basic example from the documentation (https://docs.llm-chain.xyz/docs/getting-started-tutorial/generating-your-first-llm-output):
```
Updating crates.io index
   Compiling llm-chain-openai v0.12.2
error[E0308]: mismatched types
   --> /Users/jessiewilkins/.cargo/registry/src/index.crates.io-6f17d22bba15001f/llm-chain-openai-0.12.2/src/chatgpt/executor.rs:113:60
    |
113 |         let tokens_used = num_tokens_from_messages(&model, &messages)
    |                           ------------------------         ^^^^^^^^^ expected `&[ChatCompletionRequestMessage]`, found `&Vec<ChatCompletionRequestMessage>`
    |                           |
    |                           arguments to this function are incorrect
    |
    = note: expected reference `&[async_openai::types::types::ChatCompletionRequestMessage]`
               found reference `&Vec<async_openai::types::ChatCompletionRequestMessage>`
note: function defined here
   --> /Users/jessiewilkins/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tiktoken-rs-0.4.5/src/api.rs:358:12
    |
358 |     pub fn num_tokens_from_messages(
    |            ^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
```

--- a/crates/llm-chain-openai/src/chatgpt/executor.rs
+++ b/crates/llm-chain-openai/src/chatgpt/executor.rs
@@ -8,7 +8,7 @@ use llm_chain::tokens::TokenCollection;
 
 use super::prompt::create_chat_completion_request;
 use super::prompt::format_chat_messages;
-use async_openai::error::OpenAIError;
+use async_openai::{error::OpenAIError, types::ChatCompletionRequestMessage};
 use llm_chain::prompt::Prompt;
 
 use llm_chain::tokens::PromptTokensError;
@@ -19,7 +19,7 @@ use llm_chain::traits::{ExecutorCreationError, ExecutorError};
 use async_trait::async_trait;
 use llm_chain::tokens::TokenCount;
 
-use tiktoken_rs::async_openai::num_tokens_from_messages;
+use tiktoken_rs::get_chat_completion_max_tokens;
 
 use std::sync::Arc;
 
@@ -109,9 +109,11 @@ impl traits::Executor for Executor {
     ) -> Result<TokenCount, PromptTokensError> {
         let opts_cas = self.cascade(Some(opts));
         let model = self.get_model_from_invocation_options(&opts_cas);
-        let messages = format_chat_messages(prompt.to_chat())?;
-        let tokens_used = num_tokens_from_messages(&model, &messages)
-            .map_err(|_| PromptTokensError::NotAvailable)?;
+        let messages: Vec<ChatCompletionRequestMessage> = format_chat_messages(prompt.to_chat())?;
+
+        let tokens_used =
+            get_chat_completion_max_tokens(&model, as_tiktoken_messages(messages).as_slice())
+                .map_err(|_| PromptTokensError::NotAvailable)?;
 
         Ok(TokenCount::new(
             self.max_tokens_allowed(opts),
@@ -134,6 +136,22 @@ impl traits::Executor for Executor {
     fn get_tokenizer(&self, options: &Options) -> Result<OpenAITokenizer, TokenizerError> {
         Ok(OpenAITokenizer::new(self.cascade(Some(options))))
     }
+}
+
+fn as_tiktoken_message(
+    message: &ChatCompletionRequestMessage,
+) -> tiktoken_rs::ChatCompletionRequestMessage {
+    tiktoken_rs::ChatCompletionRequestMessage {
+        role: message.role.to_string(),
+        content: message.content.clone(),
+        name: message.name.clone(),
+    }
+}
+
+fn as_tiktoken_messages(
+    messages: Vec<ChatCompletionRequestMessage>,
+) -> Vec<tiktoken_rs::ChatCompletionRequestMessage> {
+    messages.iter().map(|x| as_tiktoken_message(x)).collect()
 }
 
 pub struct OpenAITokenizer {

--- a/crates/llm-chain-openai/src/chatgpt/executor.rs
+++ b/crates/llm-chain-openai/src/chatgpt/executor.rs
@@ -110,8 +110,10 @@ impl traits::Executor for Executor {
         let opts_cas = self.cascade(Some(opts));
         let model = self.get_model_from_invocation_options(&opts_cas);
         let messages: Vec<ChatCompletionRequestMessage> = format_chat_messages(prompt.to_chat())?;
-
+        let no_messages: Vec<tiktoken_rs::ChatCompletionRequestMessage> = Vec::new();
         let tokens_used =
+            get_chat_completion_max_tokens(&model, no_messages.as_slice())
+                .map_err(|_| PromptTokensError::NotAvailable)? - 
             get_chat_completion_max_tokens(&model, as_tiktoken_messages(messages).as_slice())
                 .map_err(|_| PromptTokensError::NotAvailable)?;
 

--- a/crates/llm-chain-openai/src/chatgpt/executor.rs
+++ b/crates/llm-chain-openai/src/chatgpt/executor.rs
@@ -111,10 +111,9 @@ impl traits::Executor for Executor {
         let model = self.get_model_from_invocation_options(&opts_cas);
         let messages: Vec<ChatCompletionRequestMessage> = format_chat_messages(prompt.to_chat())?;
         let no_messages: Vec<tiktoken_rs::ChatCompletionRequestMessage> = Vec::new();
-        let tokens_used =
-            get_chat_completion_max_tokens(&model, no_messages.as_slice())
-                .map_err(|_| PromptTokensError::NotAvailable)? - 
-            get_chat_completion_max_tokens(&model, as_tiktoken_messages(messages).as_slice())
+        let tokens_used = get_chat_completion_max_tokens(&model, no_messages.as_slice())
+            .map_err(|_| PromptTokensError::NotAvailable)?
+            - get_chat_completion_max_tokens(&model, as_tiktoken_messages(messages).as_slice())
                 .map_err(|_| PromptTokensError::NotAvailable)?;
 
         Ok(TokenCount::new(

